core/modules/contextual/js/contextual.js

```diff
@@ -37,8 +37,46 @@ Backdrop.behaviors.contextualLinks = {
       // Prepend the trigger.
       $wrapper.prepend($trigger);
 
-      // Prevent child contextual links overlapping their parent ones.
-      $wrapper.siblings().find('.contextual-links-wrapper').eq(0).css('margin-top', '20px');
+      // Prevent child contextual link triggers from overlapping their parent
+      // ones by shifting their position.
+      var $child = $wrapper.siblings().find('.contextual-links-wrapper').eq(0);
+      if ($child.length > 0) {
+        var parentHeight = $child.parent().height();
+        var parentOffset = $wrapper.offset();
+        var childOffset = $child.offset();
+        var dir = $('html').attr('dir');
+        if (childOffset.top < parentOffset.top + 20) {
+          if (dir == 'ltr') {
+            if (childOffset.left > parentOffset.left - 25) {
+              // There's a collision at right, so we need to shift the child.
+              if (parentHeight >= 40) {
+                  // If there's enought vertical room in the child contextual links
+                  // region, shift the child links down.
+                $child.css('margin-top', '20px');
+              }
+              else {
+                // Otherwise shift the child links to the left.
+                $child.css('margin-right', '25px');
+              }
+            }
+          }
+          else {
+            // Handle RTL.
+            if (childOffset.left < parentOffset.left + 25) {
+              // There's a collision at left, so we need to shift the child.
+              if (parentHeight >= 40) {
+                  // If there's enought vertical room in the child contextual links
```

Comment:
minor typo here. `enought`

Reply:
Thanks, fixed (in two places). Also rewrapped the comments, which were over 80 char long.